// itch/bin/itch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
   value: true
});
var strictEquals = function strictEquals(seed, candidate) {
   return seed === candidate;
};
var arrayMatcher = function arrayMatcher(matcher) {
   return function (seed, candidates) {
      return candidates.some(function (c) {
         return matcher(seed, c);
      });
   };
};

var done = function done(result) {
   var match = function match() {
      return {
         then: function then() {
            return done(result);
         },
         evaluate: function evaluate() {
            return done(result);
         }
      };
   };
   return {
      match: match,
      matchOneOf: match,
      scratch: function scratch() {
         return result;
      },
      scratchEvaluate: function scratchEvaluate() {
         return result;
      },
      using: function using() {
         return done(result);
      }
   };
};

var notDone = function notDone(seed) {
   var matcher = arguments.length <= 1 || arguments[1] === undefined ? strictEquals : arguments[1];

   var match = function match(candidate) {
      var newMatcher = arguments.length <= 1 || arguments[1] === undefined ? matcher : arguments[1];
      return {
         then: newMatcher(seed, candidate) ? function (result) {
            return done(result);
         } : function () {
            return notDone(seed, matcher);
         },
         evaluate: newMatcher(seed, candidate) ? function (result) {
            return done(typeof result === 'function' ? result() : result);
         } : function () {
            return notDone(seed, matcher);
         }
      };
   };
   return {
      match: match,
      matchOneOf: function matchOneOf(candidates) {
         return match(candidates, arrayMatcher(matcher));
      },
      scratch: function scratch(a) {
         return a;
      },
      scratchEvaluate: function scratchEvaluate(a) {
         return typeof a === 'function' ? a() : a;
      },
      using: function using(newMatcher) {
         return notDone(seed, newMatcher);
      }
   };
};

exports.default = function (value) {
   return notDone(value);
};
